
# 🧮 NumPy Mega Homework

This workbook contains **lots** of NumPy practice tasks, grouped by topic.  
For each task, write your solution in the empty code cell that follows it.

**Rules**
- Use **NumPy** (`import numpy as np`) for array operations.
- Do **not** use Python lists for computations unless explicitly asked.
- Add comments to explain key steps.

**Sections**
1. Array Basics & Creation (10 tasks)  
2. Indexing & Slicing (8 tasks)  
3. Array Operations & Broadcasting (8 tasks)  
4. Mathematical Functions (6 tasks)  
5. Aggregations & Descriptive Stats (6 tasks)  
6. Random Numbers & Simulations (8 tasks)  
7. Linear Algebra (6 tasks)  
8. Reshaping, Stacking & Splitting (5 tasks)  
9. Real‑World Mini Case Studies (7 tasks)


## 1) Array Basics & Creation

**1.1** Import NumPy as `np` and print the installed version.

In [67]:
import numpy as np

**1.2** Create a 1D array with values from 0 to 20 (inclusive).

In [68]:
# Burada hər iki ədədi yazdırmaq üçün mütləq (0,21) qeyd etməliyik.

np.arange(0,21)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20])

In [69]:
# Bundan əlavə iki rəqəm arasındakı məsafəni biliriksə, onda "linspace"
# funksiyasını da qeyd etmək olar.
# Tipə "int" verərək tam ədədlər qaytarmasını təmin edirik.

np.linspace(0,20,21,dtype=int) 

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20])

**1.3** Create a 1D array of the first 15 even numbers.

In [70]:
np.arange(0,30,2)

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28])

**1.4** Create a 3×3 array filled with 7s (use `np.full`).

In [71]:
np.full((3,3),7)

array([[7, 7, 7],
       [7, 7, 7],
       [7, 7, 7]])

**1.5** Create a 5×5 identity matrix.

In [72]:
np.eye(5,dtype=int)

array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1]])

**1.6** Create an array of 10 numbers linearly spaced between 0 and 1 (inclusive).

In [73]:
np.linspace(0,1,10)

array([0.        , 0.11111111, 0.22222222, 0.33333333, 0.44444444,
       0.55555556, 0.66666667, 0.77777778, 0.88888889, 1.        ])

**1.7** Create an array of 12 values logarithmically spaced between 10^0 and 10^3.

In [74]:
np.logspace(0,3,12)

array([   1.        ,    1.87381742,    3.51119173,    6.57933225,
         12.32846739,   23.101297  ,   43.28761281,   81.11308308,
        151.9911083 ,  284.80358684,  533.66992312, 1000.        ])

**1.8** Create a boolean array of length 12 that alternates `True, False, True, False, ...`.

In [75]:
np.tile([True, False], 6)

array([ True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False])

**1.9** Create a 4×4 array of random integers from 10 to 99 (inclusive).

In [76]:
np.random.randint(10,100,size=(4,4))

array([[87, 22, 67, 59],
       [71, 94, 94, 95],
       [27, 22, 59, 69],
       [64, 35, 40, 95]], dtype=int32)

**1.10** Given `a = np.array([1, 2, 3], dtype=np.int32)`, change its dtype to float64 without copying if possible.

In [77]:
a=np.array([1,2,3], dtype=np.int32)
b=a.astype(np.float64)
b

array([1., 2., 3.])

## 2) Indexing & Slicing

**2.1** Create an array `x = np.arange(1, 26).reshape(5,5)`. Extract the **center 3×3 block**.

In [78]:
x=np.arange(1, 26).reshape(5,5)
x

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

In [79]:
# Answer
x[1:4, 1:4]

array([[ 7,  8,  9],
       [12, 13, 14],
       [17, 18, 19]])

**2.2** From `x` above, extract the **border elements** (top row, bottom row, leftmost column, rightmost column) into a 1D array.

In [80]:
# Answer
mask=np.ones((5,5), dtype=bool)
mask[1:4, 1:4]=False
border_elements=x[mask]
border_elements

array([ 1,  2,  3,  4,  5,  6, 10, 11, 15, 16, 20, 21, 22, 23, 24, 25])

In [81]:
# Əgər istəyiriksə ki, matrisin orta elementləri sıfır olsun, onda aşağıdakıları edirik:

x[1:4, 1:4]=0
border_elements

array([ 1,  2,  3,  4,  5,  6, 10, 11, 15, 16, 20, 21, 22, 23, 24, 25])

**2.3** From `x`, select every other row and every other column.

In [82]:
# Answer
x=np.arange(1, 26).reshape(5,5)

x[0:, 1]

array([ 2,  7, 12, 17, 22])

**2.4** Given `y = np.arange(10)`, reverse it using slicing only.

In [83]:
y=np.arange(10)
y

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [84]:
# Answer
reversed_y=y[::-1]
reversed_y

array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

**2.5** From `y`, extract elements with **odd indices** (1,3,5,...) using slicing.

In [85]:
# Answer
y[0::2]

array([0, 2, 4, 6, 8])

**2.6** Use **fancy indexing** to pick elements at indices `[0, 3, 7]` from `y`.

In [86]:
# Answer
answer=y[[0,3,7]]
answer

array([0, 3, 7])

**2.7** Given `z = np.array([3, -1, 7, 0, -5, 2])`, replace **negative** values with `0` using boolean indexing.

In [87]:
z=np.array([3,-1,7,0,-5,2])
z

array([ 3, -1,  7,  0, -5,  2])

In [88]:
# Answer
z[z<0]=0
z

array([3, 0, 7, 0, 0, 2])

**2.8** Given `M = np.arange(1,17).reshape(4,4)`, set the **diagonal** to `-1` without a loop.

In [89]:
M=np.arange(1,17).reshape(4,4)
M

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [90]:
# Answer
diagonal=np.diag_indices_from(M)
M[diagonal]=-1
M

array([[-1,  2,  3,  4],
       [ 5, -1,  7,  8],
       [ 9, 10, -1, 12],
       [13, 14, 15, -1]])

## 3) Array Operations & Broadcasting

**3.1** Compute element‑wise addition, subtraction, multiplication, and division of arrays `a=[1,2,3,4]` and `b=[10,20,30,40]`.

In [91]:
# Answer

a=np.array([1,2,3,4])
b=np.array([10,20,30,40])

print(a+b)

print((a-b))

print(a*b)

print(a/b)

[11 22 33 44]
[ -9 -18 -27 -36]
[ 10  40  90 160]
[0.1 0.1 0.1 0.1]


**3.2** Square all elements of `a` without using a Python loop.

In [92]:
print(a**2)

[ 1  4  9 16]


**3.3** Let `A = np.ones((3,3))` and `v = np.array([1,2,3])`. Add `v` to each **row** of `A` using broadcasting.

In [93]:
# Answer
A=np.ones((3,3))
v=np.array([1,2,3])

result=A+v
result

array([[2., 3., 4.],
       [2., 3., 4.],
       [2., 3., 4.]])

**3.4** Multiply each **column** of `A` by `[1,2,3]` using broadcasting.

In [94]:
# Answer
Mltp=A*v[:,np.newaxis]
Mltp

array([[1., 1., 1.],
       [2., 2., 2.],
       [3., 3., 3.]])

**3.5** Normalize array `p = np.array([3.0, 4.0, 0.0, 8.0])` to unit length (divide by its Euclidean norm).

In [95]:
# Answer
p=np.array([3.0, 4.0, 0.0, 8.0])

lenght=np.sqrt(np.sum(p**2))
normalized=p/lenght

print(normalized)
print("New lenght:", np.sqrt(np.sum(normalized**2)))

[0.31799936 0.42399915 0.         0.8479983 ]
New lenght: 1.0


**3.6** Compute the pairwise sum of vectors in `X = np.array([[1,2],[3,4],[5,6]])` with `y = np.array([10,20])` via broadcasting (result shape 3×2).

In [96]:
X=np.array([[1,2],
            [3,4],
            [5,6]])
y=np.array([10,20])

result=X+y
print("Nəticə:\n", result)

Nəticə:
 [[11 22]
 [13 24]
 [15 26]]


**3.7** Given `B = np.arange(12).reshape(3,4)`, subtract the **column means** from each column (column‑wise centering) using broadcasting.

In [97]:
# Answer
B=np.arange(12).reshape(3,4)
column_means=B.mean(axis=0)
B_centered=B-column_means
print("Centralized matrix:\n", B_centered)

Centralized matrix:
 [[-4. -4. -4. -4.]
 [ 0.  0.  0.  0.]
 [ 4.  4.  4.  4.]]


**3.8** Clip array values: given `w = np.array([-3, -1, 0, 2, 9])`, clip to the range `[0, 5]`.

In [98]:
# Answer
w=np.array([-3,-1,0,2,9])
w_clip=np.clip(w,0,5)
print(w_clip)

[0 0 0 2 5]


## 4) Mathematical Functions

**4.1** For `q = np.linspace(-2*np.pi, 2*np.pi, 50)`, compute `sin(q)`, `cos(q)`, and `tan(q)` arrays.

In [99]:
q=np.linspace(-2*np.pi, 2*np.pi, 50)
print("Sin(q):\n", np.sin(q))
print("cos(q):\n", np.cos(q))
print("Tan(q):\n", np.tan(q))

Sin(q):
 [ 2.44929360e-16  2.53654584e-01  4.90717552e-01  6.95682551e-01
  8.55142763e-01  9.58667853e-01  9.99486216e-01  9.74927912e-01
  8.86599306e-01  7.40277997e-01  5.45534901e-01  3.15108218e-01
  6.40702200e-02 -1.91158629e-01 -4.33883739e-01 -6.48228395e-01
 -8.20172255e-01 -9.38468422e-01 -9.95379113e-01 -9.87181783e-01
 -9.14412623e-01 -7.81831482e-01 -5.98110530e-01 -3.75267005e-01
 -1.27877162e-01  1.27877162e-01  3.75267005e-01  5.98110530e-01
  7.81831482e-01  9.14412623e-01  9.87181783e-01  9.95379113e-01
  9.38468422e-01  8.20172255e-01  6.48228395e-01  4.33883739e-01
  1.91158629e-01 -6.40702200e-02 -3.15108218e-01 -5.45534901e-01
 -7.40277997e-01 -8.86599306e-01 -9.74927912e-01 -9.99486216e-01
 -9.58667853e-01 -8.55142763e-01 -6.95682551e-01 -4.90717552e-01
 -2.53654584e-01 -2.44929360e-16]
cos(q):
 [ 1.          0.96729486  0.8713187   0.71834935  0.51839257  0.28452759
  0.03205158 -0.22252093 -0.46253829 -0.67230089 -0.8380881  -0.94905575
 -0.99794539 -0.981559

**4.2** Compute element‑wise `exp`, `log`, and `sqrt` for `r = np.array([1, e, e**2, 10.0])` (use `np.e` where needed).

In [100]:
r=np.array([1,np.e,np.e**2,10.0])

print("Exp:\n", np.exp(r))
print("Log:\n", np.log(r))
print("Sqrt:\n", np.sqrt(r))

Exp:
 [2.71828183e+00 1.51542622e+01 1.61817799e+03 2.20264658e+04]
Log:
 [0.         1.         2.         2.30258509]
Sqrt:
 [1.         1.64872127 2.71828183 3.16227766]


**4.3** Compute `np.floor`, `np.ceil`, and `np.round` for `s = np.array([1.2, 3.7, -2.3, -4.8])`.

In [101]:
s=np.array([1.2, 3.7, -2.3, -4.8])
print("Floor:\n", np.floor(s))
print("Ceil:\n", np.ceil(s))
print("Round:\n", np.round(s))

Floor:
 [ 1.  3. -3. -5.]
Ceil:
 [ 2.  4. -2. -4.]
Round:
 [ 1.  4. -2. -5.]


**4.4** Safely compute `log1p(x)` for `x = np.array([0, 1e-10, 1e-5, 0.1])`. Why is `log1p` preferred here? (Answer in a comment)

In [102]:
x=np.array([0,1e-10, 1e-5, 0.1])
# Bu funksiya riyazi olaraq ln(1+x)-ı hesablayır. Bu, özünəməxsus alqoritimlə 
# dəqiqliyin itməməsi üçündür. Daha detallı aşağıdakı kodlarda görəcəyik:

# Normal üsulla hesablasaq:
bad_way=np.log(1+x)

# log1p(x) üsulu ilə, daha dəqiq olacaq:
good_way=np.log1p(x)

print("Normal üsul:\n", bad_way)
print("log1p(x):\n", good_way)
# Cavaba baxaraq, yazılanları daha yaxşı görürük.

Normal üsul:
 [0.00000000e+00 1.00000008e-10 9.99995000e-06 9.53101798e-02]
log1p(x):
 [0.00000000e+00 1.00000000e-10 9.99995000e-06 9.53101798e-02]


**4.5** Compute the sigmoid function `σ(x) = 1 / (1 + exp(-x))` for `x = np.linspace(-8, 8, 9)`.

In [103]:
# Siqmoid funksiyası hər hansı bir ədədi götürüb onu 0və 1 arasında sıxışdırır.

x=np.linspace(-8,8,9)
sigmoid_x=1/(1+np.exp(-x))
sigmoid_x

array([3.35350130e-04, 2.47262316e-03, 1.79862100e-02, 1.19202922e-01,
       5.00000000e-01, 8.80797078e-01, 9.82013790e-01, 9.97527377e-01,
       9.99664650e-01])

**4.6** Standardize `t` (z‑score): given `t = np.array([10, 12, 9, 13, 11, 8])`, transform to mean 0 and std 1.

In [104]:
t=np.array([10,12,9,13,11,8])
mean_t=t.mean()
std_t=t.std()

z_score=(t-mean_t)/std_t
z_score

array([-0.29277002,  0.87831007, -0.87831007,  1.46385011,  0.29277002,
       -1.46385011])

## 5) Aggregations & Descriptive Stats

**5.1** Given `u = np.array([5, 5, 9, 1, 1, 9, 9])`, compute: mean, median, variance, std, min, max.

In [105]:
u=np.array([5,5,9,1,1,9,9])

print("Mean:",np.mean(u))
print("Median:",np.median(u))
print("Variance:",np.var(u))
print("Std:",np.std(u))
print("Min:",np.min(u))
print("Max:",np.max(u))

Mean: 5.571428571428571
Median: 5.0
Variance: 11.10204081632653
Std: 3.3319725113401715
Min: 1
Max: 9


**5.2** Compute `argmin` and `argmax` for `u`. What values do they point to? (Print index & value)

In [106]:
min_index=np.argmin(u)
max_index=np.argmax(u)

print(f"Min index:{min_index}, Value:{u[min_index]}")
print(f"Max index:{max_index}, Value:{u[max_index]}")

Min index:3, Value:1
Max index:2, Value:9


**5.3** For `V = np.arange(12).reshape(3,4)`, compute **row sums** and **column means**.

In [107]:
V=np.arange(12).reshape(3,4)
V

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [108]:
row_sums=V.sum(axis=1)
column_means=V.mean(axis=0)
print(row_sums)
print(column_means)

[ 6 22 38]
[4. 5. 6. 7.]


**5.4** For `V`, compute the cumulative sum along axis 1 and the cumulative product along axis 0.

In [109]:
cumsum_row=V.cumsum(axis=1)
cumprod_col=V.cumprod(axis=0)
print("Row-wise cumulative sum:\n", cumsum_row)
print("Column_wise cumulative product:\n", cumprod_col)

Row-wise cumulative sum:
 [[ 0  1  3  6]
 [ 4  9 15 22]
 [ 8 17 27 38]]
Column_wise cumulative product:
 [[  0   1   2   3]
 [  0   5  12  21]
 [  0  45 120 231]]


**5.5** For `u`, return the **unique** values and their **counts** (hint: `np.unique(..., return_counts=True)`).

In [110]:
values, counts=np.unique(u, return_counts=True)
print("Unique values:", values)
print("Counts:", counts)

Unique values: [1 5 9]
Counts: [2 2 3]


**5.6** Compute the **50th percentile (median)** and **90th percentile** of `u` using `np.percentile`.

In [111]:
p50=np.percentile(u, 50)
p90=np.percentile(u,90)
print("50th percentile (Median):",p50)
print("90th percentile:",p90)

50th percentile (Median): 5.0
90th percentile: 9.0


## 6) Random Numbers & Simulations


**6.1** Set a random seed (e.g., 123). Generate 10 random numbers from the uniform distribution [0,1).

In [112]:
rng=np.random.default_rng(seed=123)
random_numbers=rng.random(10)
print(random_numbers)

[0.68235186 0.05382102 0.22035987 0.18437181 0.1759059  0.81209451
 0.923345   0.2765744  0.81975456 0.88989269]


**6.2** Generate a 3×3 matrix of random integers between 1 and 100 (inclusive).

In [113]:
rng=np.random.default_rng()
matrix=rng.integers(low=1,high=101, size=(3,3))
print(matrix)

[[67 41 84]
 [97 57 79]
 [12 53 64]]


In [114]:
# Başqa bir yol kimi bunu da göstərə bilərik:
random_matrix=np.random.randint(1,101,size=(3,3))
print(random_matrix)

[[ 29 100  47]
 [ 93  78  27]
 [  1  85  54]]


**6.3** Draw 1,000 samples from a standard normal distribution and compute sample mean and std. Compare to theoretical values.

In [115]:
rng=np.random.default_rng(seed=42)
samples=rng.standard_normal(1000)
sample_mean=samples.mean()
sample_std=samples.std()
print(f"Sample mean:{sample_mean:.4f} (Theoretical: {0:.4f})")
print(f"Sample std:{sample_std:.4f} (Theoretical: {1:.4f})")

Sample mean:-0.0289 (Theoretical: 0.0000)
Sample std:0.9887 (Theoretical: 1.0000)


**6.4** Simulate 1,000 fair coin flips (use 0/1). Estimate the probability of heads.

In [116]:
# rng-random number generator: təsadüfi ədədlər yaratmaq üçün bu sətri yazırıq:
rng=np.random.default_rng(seed=42)
# high=2 yazdıq ki, konkret 0 və 1 ədədlərini mənimsəsin.
flips=rng.integers(low=0, high=2, size=1000)
# flips.mean() yazılması ehtimalın tapılması üçündür.
prob_heads=flips.mean()
print(f"Estimeted probability of heads: {prob_heads: .4f} (Theoretical: 0.5000)")

Estimeted probability of heads:  0.5090 (Theoretical: 0.5000)


**6.5** Simulate rolling two fair dice 10,000 times. Estimate the probability that the sum equals 7.

In [117]:
rng=np.random.default_rng(seed=123)

# Burada hər iki zər üçün 10000 atış yaradırıq:
dice1=rng.integers(low=1, high=7, size=10000)
dice2=rng.integers(low=1, high=7, size=10000)

# Buradan alınan nəticələri toplayırıq:
total_sums=dice1+dice2
# Ehtimal tapılır:
prob_seven=np.mean(total_sums==7)
print(f"Estimeted probability: {prob_seven:.4f} (Theoretical:{1/6: .4f})")

Estimeted probability: 0.1732 (Theoretical: 0.1667)


**6.6** Draw 500 samples from a binomial distribution (n=20, p=0.4). Compute mean and variance; compare to theory.

In [118]:
rng=np.random.default_rng(seed=123)
sample=rng.binomial(n=20, p=0.4, size=500)
sample_mean=np.mean(samples)
sample_var=np.var(sample)

print(f" Sample mean: {sample_mean: .4f} (Theoretical: {8.0: .4f}")
print(f" Sample var: {sample_var: .4f} (Theoretical: {4.8: .4f}")

 Sample mean: -0.0289 (Theoretical:  8.0000
 Sample var:  4.8128 (Theoretical:  4.8000


**6.7** Use `np.random.multivariate_normal` to draw 1,000 2D samples with mean `[0,0]` and covariance `[[1,0.5],[0.5,1]]`. Compute the sample covariance matrix.

In [119]:
rng=np.random.default_rng(seed=123)

theoretical_mean=[0,0]
theoretical_cov=[[1,0.5], [0.5,1]]

samples=rng.multivariate_normal(mean=theoretical_mean, cov=theoretical_cov, size=1000)
sample_cov_matrix=np.cov(samples, rowvar=False)

print("Theoretical Covariance Matrix:")
print(np.array(theoretical_cov))

print("\nSample Covariance Matrix")
print(sample_cov_matrix)

Theoretical Covariance Matrix:
[[1.  0.5]
 [0.5 1. ]]

Sample Covariance Matrix
[[1.06626812 0.51805206]
 [0.51805206 1.0067271 ]]


**6.8** Monte Carlo: Estimate π by randomly sampling 100,000 points in the unit square and counting how many fall inside the unit quarter‑circle.

In [120]:
rng=np.random.default_rng(seed=123)

x=rng.random(100000)
y=rng.random(100000)

distance_squared= x**2 + y**2

inside_circle_ratio=np.mean(distance_squared <=1)
pi=inside_circle_ratio*4
print(f"Estimated pi: {pi: .4f} (Real pi:{np.pi: .4f})")

Estimated pi:  3.1436 (Real pi: 3.1416)


## 7) Linear Algebra

**7.1** Compute the dot product of `a = [1,2,3]` and `b = [4,5,6]`.

In [121]:
a=np.array([1,2,3])
b=np.array([4,5,6])

result=np.dot(a,b)
print(f"Dot product: {result}")

Dot product: 32


**7.2** Given `A = [[2,1],[1,2]]`, compute determinant, inverse, and verify `A @ A_inv ≈ I`.

In [124]:
A=np.array([[2,1], [1,2]])
det_A=np.linalg.det(A)
A_inv=np.linalg.inv(A)
is_identity=np.allclose(A@A_inv, np.eye(2))
print(f"Determinant: {det_A: .4f}")
print("\nInverse Matrix:")
print(A_inv)
print(f"\nIs it Identity Matrix? {is_identity}")

Determinant:  3.0000

Inverse Matrix:
[[ 0.66666667 -0.33333333]
 [-0.33333333  0.66666667]]

Is it Identity Matrix? True


**7.3** For `A` above, compute eigenvalues and eigenvectors, and verify `A @ v ≈ λ v`.

In [125]:
eigenvalues, eigenvectors=np.linalg.eig(A)

lam0=eigenvalues[0]
v0=eigenvectors[:,0]

left_side=A@v0
right_side=lam0*v0
is_verified=np.allclose(left_side, right_side)

print(f"Eigenvalues: {eigenvalues}")
print("\nEigenvectors Matrix:")
print(eigenvectors)
print(f"\nDoes A@v0 ≈ λ0*v0? {is_verified}")

Eigenvalues: [3. 1.]

Eigenvectors Matrix:
[[ 0.70710678 -0.70710678]
 [ 0.70710678  0.70710678]]

Does A@v0 ≈ λ0*v0? True


**7.4** Solve the linear system `3x + 2y = 7` and `2x + 5y = 8`.

In [127]:
A=np.array([[3,2], [2,5]])
B=np.array([7,8])

# Sistemi həll etmək üçün numpy kitabxanasının konkret bir funksiyası var:
solution=np.linalg.solve(A,B)

# Nəticələri massivdən fərdi dəyişənlərə ayırırıq:
x=solution[0]
y=solution[1]

print(x, y)

# Formatlama metodu ilə ədədləri daha oxunaqlı vəziyyətə gətiririk:
print(f"Solution: x={x: .4f}, y={y: .4f}")

1.727272727272727 0.9090909090909091
Solution: x= 1.7273, y= 0.9091


**7.5** Create a random 4×4 matrix and compute its **rank**.

In [130]:
matrix=np.random.rand(4,4)

rank=np.linalg.matrix_rank(matrix)

print(matrix)
print("Matrix's rank:", rank)

[[0.69251586 0.27071589 0.99139731 0.32707402]
 [0.8459757  0.64654911 0.5234439  0.89140678]
 [0.50219853 0.1527002  0.74586277 0.41517452]
 [0.01029606 0.12396773 0.21083506 0.76592521]]
Matrix's rank: 4


**7.6** Perform QR decomposition of a random 4×3 matrix and verify `Q @ R ≈ A`.

In [133]:
# Generatoru yaradırıq:
rng=np.random.default_rng(seed=123)
A=rng.random((4,3))
Q,R=np.linalg.qr(A)
is_verified=np.allclose(Q@R,A)
print("Original matrix A:\n", A)
print("\Orthogonal matrix Q:\n", Q)
print("\nUpper Triangular matrix R:\n", R)

print(f"Does Q@R approximate A? {is_verified}")

Original matrix A:
 [[0.68235186 0.05382102 0.22035987]
 [0.18437181 0.1759059  0.81209451]
 [0.923345   0.2765744  0.81975456]
 [0.88989269 0.51297046 0.2449646 ]]
\Orthogonal matrix Q:
 [[-0.46600298  0.65287377  0.11318618]
 [-0.12591424 -0.36457431 -0.79963306]
 [-0.63058599  0.20038828 -0.36805465]
 [-0.60774019 -0.63299759  0.46077287]]

Upper Triangular matrix R:
 [[-1.46426501 -0.53338651 -0.87074319]
 [ 0.         -0.29827923 -0.1429944 ]
 [ 0.          0.         -0.81327735]]
Does Q@R approximate A? True


## 8) Reshaping, Stacking & Splitting

**8.1** Create `a = np.arange(24)` and reshape to (4,6).

In [136]:
a=np.arange(24)
reshaped_a=a.reshape(4,6)
print(a)
print("\n", reshaped_a)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23]

 [[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]]


**8.2** From the (4,6) array, extract a (2,3) block from the center.

In [138]:
center_block=reshaped_a[1:3, 2:5]
print(center_block)

[[ 8  9 10]
 [14 15 16]]


**8.3** Vertically stack two arrays of shape (2,3) to obtain (4,3); then horizontally stack to obtain (2,6).

In [141]:
x=np.array([[1,2,3],
            [4,5,6]])
y=np.array([[7,8,9],
            [10,11,12]])

v_stacked=np.vstack((x,y))
h_stacked=np.hstack((x,y))

print(v_stacked)
print("\n",h_stacked)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]

 [[ 1  2  3  7  8  9]
 [ 4  5  6 10 11 12]]


**8.4** Split an array of length 20 into 4 equal parts; then split into parts of lengths `[3,5,12]`.

In [149]:
# İlk olaraq massivi yaradaq və 4 bərabər hissəyə bölək:
arr=np.arange(20)
hisse=np.split(arr,4)

# İndi xüsusi hissələrə böməni edirik:
hisse_xususi=np.split(arr, [3,8])

print(hisse[0])
print(hisse[1])
print(hisse[2])
print(hisse[3])

print("\n Ferqli hisse")
print(hisse_xususi[0])
print(hisse_xususi[1])
print(hisse_xususi[2])

[0 1 2 3 4]
[5 6 7 8 9]
[10 11 12 13 14]
[15 16 17 18 19]

 Ferqli hisse
[0 1 2]
[3 4 5 6 7]
[ 8  9 10 11 12 13 14 15 16 17 18 19]


**8.5** Use `np.concatenate` to combine three arrays along axis 0 and along axis 1 (where shapes allow).

In [154]:
massiv1=np.array([[1,2],
                  [3,4]])
massiv2=np.array([[5,6],
                  [7,8]])
massiv3=np.array([[9,10],
                  [11,12]])
# Axis=0 olarsa:
axis0=np.concatenate((massiv1, massiv2, massiv3), axis=0)

# Axis=1 olarsa:
axis1=np.concatenate((massiv1, massiv2, massiv3), axis=1)

print(massiv1)
print(massiv2)
print(massiv3)

print("Üst-üstə")
print(axis0)
print("Yan-yana")
print(axis1)

[[1 2]
 [3 4]]
[[5 6]
 [7 8]]
[[ 9 10]
 [11 12]]
Üst-üstə
[[ 1  2]
 [ 3  4]
 [ 5  6]
 [ 7  8]
 [ 9 10]
 [11 12]]
Yan-yana
[[ 1  2  5  6  9 10]
 [ 3  4  7  8 11 12]]


## 9) Real‑World Mini Case Studies

**9.1 Finance**: Prices of a stock for 10 days are `p = [100, 102, 101, 103, 105, 104, 106, 107, 110, 108]`. Compute **daily returns** (percentage) and **cumulative return**.

**9.2 Demography**: Ages of 20 people are in an array. Compute mean, median, std, and the proportion aged 18–30. *(Generate random integers 10–80 with a fixed seed.)*

**9.3 Quality Control**: Simulate 1,000 measurements with true value 50 and measurement noise `N(0, 2^2)`. What fraction of measurements fall within ±3 of the true value?

**9.4 A/B Testing**: Variant A converts 48/500 users; Variant B converts 62/500. Compute conversion rates and the **absolute lift** (B−A). Simulate 10,000 bootstrap samples of the lift.

**9.5 Sports Analytics**: Given player scores over 8 games, compute moving average with window size 3 using vectorized operations (no loops). *(Generate random integers 0–30.)*

**9.6 Image Mini‑Task**: Create a 10×10 grayscale image as a NumPy array with a white square (value 255) in the center and 0 elsewhere. Compute its mean pixel value.

**9.7 Logistics**: There are 5 warehouses with inventories `[40, 55, 20, 80, 60]` and 4 stores with demands `[30, 50, 40, 35]`. Using NumPy (not loops), compute how many units remain after fulfilling demands sequentially from warehouses.